### PR TITLE
feat: ensure snowflake connection is read only

### PR DESCRIPTION
**Pull request by fontanierh**
## Description

fixes https://github.com/dust-tt/tasks/issues/1403

On creating the Snowflake connector, we want to ensure the provided `role` only has the necessary grants to query tables -- nothing else:
- we allow `SELECT` grants on tables and future tables
- we allow `USAGE` grants on DBs, Schemas and Warehouses
- we disallow any other type of grants

Also moved the `EXCLUDE_DATABASES` and `EXCLUDE_SCHEMAS` to types as we need those from `connectors`. Added "SNOWFLAKE_SAMPLE_DATA" as an excluded DB (not useful, I think every / a lot of snowflake DB have it)

I refactored a bit snowflake api: 
- utils now return `Result`s (avoid wrapping whole public methods in try/catch). Will also help to have typed errors in the next PR (needed to display clear error messages for the admin)
- added types and schema decoding to the public methods

## Risk

N/A

## Deploy Plan

Deploy `front` and `connectors`

### CODE REVIEW

**Review by Fraggle — APPROVED** [review]
LGTM.
Can we make it so we check the grants on every sync as well ? Or before trying to execute a query from rust.

**Comment by fontanierh** [comment]
@Fraggle very good idea, we should do on every sync and maybe delete all tables if connection is no longer RO.
on every query I'm a bit worried about latency

**Comment by Fraggle** [comment]
@fontanierh if the sync are often enough, works for me.

**Comment by fontanierh** [comment]
Changed it to every 10 minutes

File: connectors/src/connectors/snowflake/temporal/activities.ts [hunk 1]
```diff
@@ -66,6 +52,42 @@ export async function syncSnowflakeConnection(connectorId: ModelId) {
     }),
   ]);
 
+  const grantsRes = await fetchGrants({ credentials, connection });
+  if (grantsRes.isErr()) {
+    throw grantsRes.error;
+  }
+  const grantsCheck = areGrantsAreReadonly(grantsRes.value);
```
Comment (Fraggle) [inline connectors/src/connectors/snowflake/temporal/activities.ts:59]: Nit: why not make the call directly in `areGrantsAreReadonly` (also: `areGrantsAreReadonly` => `areGrantsReadonly`)

Comment (fontanierh) [reply connectors/src/connectors/snowflake/temporal/activities.ts:59]: Which call ? The fetch ?

Comment (Fraggle) [reply connectors/src/connectors/snowflake/temporal/activities.ts:59]: yep

Comment (fontanierh) [reply connectors/src/connectors/snowflake/temporal/activities.ts:59]: Yeah would make sense, we don't use the grants for anything else

Comment (Fraggle) [reply connectors/src/connectors/snowflake/temporal/activities.ts:59]: You can keep the method if you want, just make the call in the `areGrantsReadonly` method.

Comment (fontanierh) [reply connectors/src/connectors/snowflake/temporal/activities.ts:59]: I kept only one method (`isConnectionReadonly`)

**Pull request merged**